stan/math/rev/constraint/simplex_constrain.hpp

```diff
@@ -73,59 +60,42 @@ inline auto simplex_constrain(const T& y) {
  * and increment the specified log probability reference with
  * the log absolute Jacobian determinant of the transform.
  *
- * The simplex transform is defined through a centered
- * stick-breaking process.
+ * The simplex transform is defined using the inverse of the
+ * isometric log ratio (ILR) transform. This is equivalent to
+ * `softmax(sum_to_zero_constrain(y))`.
  *
  * @tparam T type of the vector to constrain
  * @param y Free vector input of dimensionality N.
  * @param lp Log probability reference to increment.
  * @return Simplex of dimensionality N + 1.
  */
 template <typename T, require_rev_col_vector_t<T>* = nullptr>
-auto simplex_constrain(const T& y, scalar_type_t<T>& lp) {
+inline auto simplex_constrain(const T& y, scalar_type_t<T>& lp) {
   using ret_type = plain_type_t<T>;
 
-  size_t N = y.size();
+  const auto N = y.size();
   arena_t<T> arena_y = y;
-  arena_t<Eigen::VectorXd> arena_z(N);
-  Eigen::VectorXd x_val(N + 1);
-
-  double stick_len(1.0);
-  for (Eigen::Index k = 0; k < N; ++k) {
-    double log_N_minus_k = std::log(N - k);
-    double adj_y_k = arena_y.val().coeff(k) - log_N_minus_k;
-    arena_z.coeffRef(k) = inv_logit(adj_y_k);
-    x_val.coeffRef(k) = stick_len * arena_z.coeff(k);
-    lp += log(stick_len);
-    lp -= log1p_exp(-adj_y_k);
-    lp -= log1p_exp(adj_y_k);
-    stick_len -= x_val(k);
-  }
-  x_val.coeffRef(N) = stick_len;
 
-  arena_t<ret_type> arena_x = x_val;
+  double lp_val = 0.0;
+  arena_t<ret_type> arena_x = simplex_constrain(arena_y.val(), lp_val);
+  lp += lp_val;
 
   if (unlikely(N == 0)) {
     return ret_type(arena_x);
   }
 
-  reverse_pass_callback([arena_y, arena_x, arena_z, lp]() mutable {
-    int N = arena_y.size();
-    double stick_len_val = arena_x.val().coeff(N);
-    double stick_len_adj = arena_x.adj().coeff(N);
-    for (Eigen::Index k = N; k-- > 0;) {
-      arena_x.adj().coeffRef(k) -= stick_len_adj;
-      stick_len_val += arena_x.val().coeff(k);
-      double log_N_minus_k = std::log(N - k);
-      double adj_y_k = arena_y.val().coeff(k) - log_N_minus_k;
-      arena_y.adj().coeffRef(k) -= lp.adj() * inv_logit(adj_y_k);
-      arena_y.adj().coeffRef(k) += lp.adj() * inv_logit(-adj_y_k);
-      stick_len_adj += lp.adj() / stick_len_val;
-      stick_len_adj += arena_x.adj().coeff(k) * arena_z.coeff(k);
-      double arena_z_adj = arena_x.adj().coeff(k) * stick_len_val;
-      arena_y.adj().coeffRef(k)
-          += arena_z_adj * arena_z.coeff(k) * (1.0 - arena_z.coeff(k));
-    }
+  reverse_pass_callback([arena_y, arena_x, lp]() mutable {
+    const auto& res_val = to_ref(arena_x.val());
+
+    // backprop for log jacobian contribution to log density
+    arena_x.adj().array() += lp.adj() / res_val.array();
```

Comment:
This is kind of confusing, why are we accumulating into the return values adjoints here?

Reply:
We can avoid it by making a copy of the adjoints if we want, though I'm pretty sure that nothing else in the AD graph can depend on the adjoints of this by this point

Reply:
Can you just use `auto lp_adj =` so then it's just an expression and you don't need any memory

Reply:
Can you elaborate?

Reply:
I mean 
```
 auto lp_adj = lp.adj() / res_val.array();
```

then in the below you can do `res_val * (arena_x.adj() + lp_adj).dot(rev_val)`